### ISSUE TITLE
tmbundle: Make tmLanguage compatible with PCRE2

### ISSUE THREAD

**Issue by lildude** [marko-js/marko-tmbundle#19]
👋 I'm the lead maintainer of the https://github.com/github/linguist library which is used for language detection and providing the syntax highlighting for languages on GitHub.com, and we use this grammar.

Our grammar compiler has found several problems with your grammar which I thought I'd let you know about.

These regexes have quite a few problems as you can see in the regex101 link after each:

https://github.com/marko-js/marko-tmbundle/blob/60ded48bea6d6eccea81e858168e20e0b1b78b01/syntaxes/marko.tmLanguage.json#L779

https://regex101.com/r/pSG73T/1

https://github.com/marko-js/marko-tmbundle/blob/60ded48bea6d6eccea81e858168e20e0b1b78b01/syntaxes/marko.tmLanguage.json#L108

... and repeated again at:

https://github.com/marko-js/marko-tmbundle/blob/60ded48bea6d6eccea81e858168e20e0b1b78b01/syntaxes/marko.tmLanguage.json#L130

https://regex101.com/r/NlVs41/1

These are the errors our compiler reported:

- Invalid regex in grammar: `text.marko` (in `syntaxes/marko.tmLanguage.json`) contains a malformed regex (regex "`(?=[,;\](]|/>|(?<=[^=])>|(?<!(?:`...": nothing to repeat (at offset 105))
- Invalid regex in grammar: `text.marko` (in `syntaxes/marko.tmLanguage.json`) contains a malformed regex (regex "`(?=[,;\]]|/>|(?<=[^=])>|(?<!(?:^`...": nothing to repeat (at offset 104))
- Invalid regex in grammar: `text.marko` (in `syntaxes/marko.tmLanguage.json`) contains a malformed regex (regex "`(?=[,;\]]|/>|(?<=[^=])>|(?<!(?:^`...": nothing to repeat (at offset 104))

**Comment by DylanPiercey** [comment]
Will look shortly. Thanks for the report!

**Comment by DylanPiercey** [comment]
@lildude from what I can tell the regex functions correctly. The validator you are using is using PCRE2 however my understanding was that the regex's in tmgrammars are intended to be handled by [oniguruma](https://github.com/kkos/oniguruma). Is this not the case for linguist?

**Comment by lildude** [comment]
> Is this not the case for linguist?

No. GitHub uses PCRE for grammar parsing for performance reasons.

**Comment by RedCMD** [comment]
VSCode uses oniguruma 
TextMate 2.0 uses Onigmo
GitHub uses PCRE (not PCRE2)
https://github.com/RedCMD/TmLanguage-Syntax-Highlighter/blob/main/documentation/README.md#regex

here's my VSCode extension to help with authoring TextMate grammars https://marketplace.visualstudio.com/items?itemName=RedCMD.tmlanguage-syntax-highlighter

**Comment by DylanPiercey** [comment]
@RedCMD already using the extension, which is great BTW!

Would be awesome if it could highlight incompatibilities 🙏

**Comment by RedCMD** [comment]
good news
I have already started implementing it
just need to install the pre-release version of the extension 

feedback is always welcome if you find any issues

**Comment by DylanPiercey** [comment]
Seems to be working great @RedCMD 🙌

Still have to figure out how I'm going to convert our gnarly (oniguruma compatible) regexp to work in PCRE though 😆

**Comment by RedCMD** [comment]
the `nothing to repeat` errors are easy
PCRE doesnt support quantifier chains `a{2}+`
so just wrap with non-capture group `(?:a{2})+`
```regex
\\+{2}*
(?:\\+{2})*
```

the `lookbehind assertion is not fixed length at offset` errors will be harder
requiring knowledge of the whole grammar
prob put some rules inside it to capture the text you don't want it to end on

you are allowed some top level non-fixed lengths
eg `(?<=a|bc)` is allowed
but not `(?<=(?:a|bc))` due to the nesting
infinity quantifiers not allowed either `(?<=a+)`


> These regexes have quite a few problems as you can see in the regex101 link after each:

@lildude you copied the JSON stringified regexes into regex101
need to de-stringify them first
